kms/keysource.go

```
@@ -162,37 +161,56 @@ func (key MasterKey) createSession() (*session.Session, error) {
 }
 
 // ToMap converts the MasterKey to a map for serialization purposes
-func (key MasterKey) ToMap() map[string]string {
-	out := make(map[string]string)
+func (key MasterKey) ToMap() map[string]interface{} {
+	out := make(map[string]interface{})
 	out["arn"] = key.Arn
 	if key.Role != "" {
 		out["role"] = key.Role
 	}
 	out["created_at"] = key.CreationDate.UTC().Format(time.RFC3339)
 	out["enc"] = key.EncryptedKey
 	if key.EncryptionContext != nil {
-		var outContexts []string
+		outcontext := make(map[string]string)
 		for k, v := range key.EncryptionContext {
-			outContexts = append(outContexts, k+":"+*v)
+			outcontext[k] = *v
 		}
-		sort.Strings(outContexts)
-		out["context"] = strings.Join(outContexts, ",")
+		out["context"] = outcontext
 	}
 	return out
 }
 
-// ParseKMSContext takes a comma-separated list of KMS context key:value pairs and returns a map
-func ParseKMSContext(in string) map[string]*string {
-	if in == "" {
-		return nil
-	}
+// ParseKMSContext takes either a KMS context map or a comma-separated list of KMS context key:value pairs and returns a map
+func ParseKMSContext(in interface{}) map[string]*string {
 	out := make(map[string]*string)
-	for _, kv := range strings.Split(in, ",") {
-		kv := strings.Split(kv, ":")
-		if len(kv) != 2 {
+
+	switch in := in.(type) {
+	case map[string]interface{}:
+		if len(in) == 0 {
+			return nil
+		}
+		for k, v := range in {
+			var v = v.(string)
+			out[k] = &v
+		}
+	case map[interface{}]interface{}:
+		if len(in) == 0 {
 			return nil
 		}
-		out[kv[0]] = &kv[1]
+		for k, v := range in {
+			var v = v.(string)
+			out[k.(string)] = &v
+		}
+	case string:
+		if in == "" {
+			return nil
+		}
+		for _, kv := range strings.Split(in, ",") {
+			kv := strings.Split(kv, ":")
+			if len(kv) != 2 {
+				return nil
```

I feel we should also show a warning here instead of silently skipping.